File: app/src/main/java/es/etologic/mahjongscoring2/app/main/old_games/OldGamesViewModel.java
package es.etologic.mahjongscoring2.app.main.old_games;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;

import java.util.List;

import es.etologic.mahjongscoring2.app.base.BaseViewModel;
import es.etologic.mahjongscoring2.domain.model.GameWithRounds;
import es.etologic.mahjongscoring2.domain.use_cases.DeleteGameUseCase;
import es.etologic.mahjongscoring2.domain.use_cases.GetGamesUseCase;
import io.reactivex.schedulers.Schedulers;

import static es.etologic.mahjongscoring2.app.model.ShowState.*;

class OldGamesViewModel extends BaseViewModel {

    //FIELDS
    private GetGamesUseCase getGamesUseCase;
    private DeleteGameUseCase deleteGameUseCase;
    private MutableLiveData<List<GameWithRounds>> allGames = new MutableLiveData<>();

    //CONSTRUCTOR
    OldGamesViewModel(GetGamesUseCase getGamesUseCase, DeleteGameUseCase deleteGameUseCase) {
        this.getGamesUseCase = getGamesUseCase;
        this.deleteGameUseCase = deleteGameUseCase;
    }

    //OBSERVABLES
    LiveData<List<GameWithRounds>> getGames() { return allGames; }

    //METHODS
    void getAllGames() {
        disposables.add(
                getGamesUseCase.getAllWithRounds()
                        .subscribeOn(Schedulers.io())
                        .doOnSubscribe(observer -> progressState.postValue(SHOW))
                        .doOnEvent((observer, throwable) -> progressState.postValue(HIDE))
                        .subscribe(allGames::postValue, error::postValue));
    }
    void deleteGame(long gameId) {
        disposables.add(
                deleteGameUseCase.deleteGame(gameId)
                        .subscribeOn(Schedulers.io())
                        .doOnSubscribe(observer -> progressState.postValue(SHOW))
                        .doOnEvent((observer, throwable) -> progressState.postValue(HIDE))
                        .subscribe(result -> getAllGames(), error::postValue));
    }
}

File: app/src/main/java/es/etologic/mahjongscoring2/app/main/old_games/OldGamesViewModel.kt
package es.etologic.mahjongscoring2.app.main.old_games

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import es.etologic.mahjongscoring2.app.base.BaseViewModel
import es.etologic.mahjongscoring2.app.model.ShowState.HIDE
import es.etologic.mahjongscoring2.app.model.ShowState.SHOW
import es.etologic.mahjongscoring2.domain.model.GameWithRounds
import es.etologic.mahjongscoring2.domain.use_cases.DeleteGameUseCase
import es.etologic.mahjongscoring2.domain.use_cases.GetGamesUseCase
import io.reactivex.schedulers.Schedulers

internal class OldGamesViewModel(
    private val getGamesUseCase: GetGamesUseCase,
    private val deleteGameUseCase: DeleteGameUseCase
) : BaseViewModel() {
    
    private val allGames = MutableLiveData<List<GameWithRounds>>()
    
    internal fun getGames(): LiveData<List<GameWithRounds>> = allGames
    
    fun deleteGame(gameId: Long) {
        disposables.add(
            deleteGameUseCase.deleteGame(gameId)
                .subscribeOn(Schedulers.io())
                .doOnSubscribe { progressState.postValue(SHOW) }
                .doOnEvent { _, _ -> progressState.postValue(HIDE) }
                .subscribe({ getAllGames() }, { error.postValue(it) })
        )
    }
    
    //METHODS
    fun getAllGames() {
        disposables.add(
            getGamesUseCase.getAllWithRounds()
                .subscribeOn(Schedulers.io())
                .doOnSubscribe { progressState.postValue(SHOW) }
                .doOnEvent { _, _ -> progressState.postValue(HIDE) }
                .subscribe(allGames::postValue) {
                    allGames.postValue(ArrayList())
                    error.postValue(it)
                }
        )
    }
}
